File: 2-add-two-numbers/add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    // Approach: Can just start adding from l1 and l2 because we add the singles place first anyways when we add. 
    // Problem: A lot of edge cases. 
    let dummy = new ListNode()
    let curr = dummy
    let carry = 0;
    while (l1 || l2 || carry) {
        const val1 = (l1) ? l1.val : 0;
        const val2 = (l2) ? l2.val : 0;

        const value = val1 + val2 + carry
        carry = Math.floor(value / 10)
        const newVal = value % 10
        curr.next = new ListNode(newVal)

        curr = curr.next
        l1 = l1 ? l1.next : null
        l2 = l2 ? l2.next : null
    }
    return dummy.next
};

File: 2-add-two-numbers/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        # inputs: l1 and l2
        # output: a linked list representing the sum of the digits
        # 1 -> 2 -> 3
        # 1 -> 2 -> 3
        # 2 -> 4 -> 6
        dummy = ListNode()
        curr = dummy
        carry = 0

        while l1 or l2 or carry:
            d1 = l1.val if l1  else 0
            d2 = l2.val if l2 else 0
            total = d1 + d2 + carry
            d3 = ListNode(total % 10)
            carry = total // 10

            curr.next = d3
            curr = curr.next

            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        
        return dummy.next


       

        
        